README.md
# Befunge Support in VS Code

[![Version](https://vsmarketplacebadge.apphb.com/version/kagof.befunge.svg)](https://marketplace.visualstudio.com/items?itemName=kagof.befunge)
[![Installs](https://vsmarketplacebadge.apphb.com/installs/kagof.befunge.svg)](https://marketplace.visualstudio.com/items?itemName=kagof.befunge)
[![Build Status](https://img.shields.io/travis/kagof/VSCode-Befunge.svg)](https://travis-ci.org/kagof/VSCode-Befunge)
[![Known Vulnerabilities](https://snyk.io/test/github/kagof/vscode-befunge/badge.svg?targetFile=package.json)](https://snyk.io/test/github/kagof/vscode-befunge?targetFile=package.json)

## Features

* Implements Syntax Highlighting in VS Code for [Befunge-93](https://esolangs.org/wiki/Befunge) and [Befunge-98](https://esolangs.org/wiki/Funge-98) programming languages.

* Adds alignment guides coming from the directional characters (`>`, `v`, `<`, `^`, `?`).

Associates `.bf`, `.befunge`, and `.b93` files with Befunge-93, and `.b98` files with Befunge-98.

## Screenshots

![FizzBuzz Example](https://raw.githubusercontent.com/kagof/VSCode-Befunge/master/assets/screenshots/screenshot-fizzbuzz.png)

![Factorial Example](https://raw.githubusercontent.com/kagof/VSCode-Befunge/master/assets/screenshots/screenshot-factorial.png)

## Installation

1. Open VS Code, select the extensions icon (Windows: `ctrl` + `shift` + `X`, Mac: `cmd` + `shift` + `X`) and search for `befunge`. Or maybe you are already here in VS Code. Nice!
2. Click `install`
3. Restart or reload VS Code when prompted
4. Add `"befunge.guides.enable": true"` to your Workspace or User Settings if you want to enable Befunge alignment guides.
    * note: it is recommended that you also add `"editor.renderIndentGuides": false` to your workspace settings if you enable this
5. (Optional) add the custom theming described in [Extension Settings](#extension-settings)

## Extension Settings

Adds the following new configuration settings:

| Name | Default (Type) | Description |
|------|----------------|-------------|
|`"befunge.guides.enabled"`|`true` (boolean)|Enable the Befunge guides coming from arrow characters.|
|`"befunge.guides.wrap"`|`false` (boolean)|Enable the Befunge guides coming from arrow characters to wrap to the other side when the edge of the code torus is hit.|
|`"befunge.guides.color.dark"`|`"#3c3c3c"` (*#rrggbb* hex string)|The Befunge guide line color to use for a dark theme.|
|`"befunge.guides.color.light"`|`"#dcdcdc"` (*#rrggbb* hex string)|The Befunge guide line color to use for a light theme.|

To get the colors shown in the [screenshots](#screenshots), some changes to your User Settings are necessary:

1. Open your user settings (Windows: `ctrl` + `,` Mac: `cmd` + `,`)
2. Add the following json to your User Settings (or Workspace Settings to only enable it on your current workspace):

``` javascript
    // custom colors
    "editor.tokenColorCustomizations": {
        "textMateRules": [

            // Befunge directional character styles
            {
                // bf93: v^><
                // bf98: v^<>x
                "scope": "keyword.control.direction.absolute.befunge",
                    "settings": {
                        "fontStyle": "bold",
                        "foreground": "#f0bf00"
                    }
            },
            {
                // bf9*: ?
                "scope": "keyword.control.direction.random.befunge",
                    "settings": {
                        "fontStyle": "bold",
                        "foreground": "#f0bf00"
                    }
            },
            {
                // bf98: []
                "scope": "keyword.control.direction.rotate.befunge",
                    "settings": {
                        "fontStyle": "bold",
                        "foreground": "#f0bf00"
                    }
            },
            {
                // bf98: r
                "scope": "keyword.control.direction.reflect.befunge",
                    "settings": {
                        "fontStyle": "bold",
                        "foreground": "#f0bf00"
                    }
            },
            {
                // bf93: |_
                // bf98: |_w
                "scope": "keyword.control.direction.conditional.befunge",
                    "settings": {
                        "fontStyle": "bold",
                        "foreground": "#f0bf00"
                    }
            },
            {
                // bf98: hlm
                "scope": "keyword.control.direction.3d.befunge",
                    "settings": {
                        "fontStyle": "bold",
                        "foreground": "#f0bf00"
                    }
            },
            // befunge end character styles
            {
                // bf93: @
                // bf98: @q
                "scope": "keyword.control.end.befunge",
                    "settings": {
                        "fontStyle": "bold",
                        "foreground": "#ff0000"
                    }
            },
            // befunge stack manipulation character styles
            {
                // bf93: :\$
                // bf98: :\$n{}u
                "scope": "keyword.operator.stack.befunge",
                    "settings": {
                        "foreground": "#b405ff"
                    }
            },
            // befunge I/O operators
            {
                // bf9*: &~
                "scope": "keyword.operator.io.input.befunge",
                    "settings": {
                        "foreground": "#128024"
                    }
            },
            {
                // bf9*: .,
                "scope": "keyword.operator.io.output.befunge",
                    "settings": {
                        "foreground": "#128024"
                    }
            },
            {
                // bf98: io
                "scope": "keyword.operator.io.file.befunge",
                    "settings": {
                        "foreground": "#128024"
                    }
            },
            {
                // bf93: pg
                // bf98: pgs
                "scope": "keyword.operator.io.code.befunge",
                    "settings": {
                        "foreground": "#128024"
                    }
            },
        ]
    }
```

As VS Code [does not (yet?) allow themes to be set on a per language basis](https://github.com/Microsoft/vscode/issues/20652), this adds to your *global* theme. However, all scopes used in this blurb end in `.befunge`, ie, this should not affect your theme anywhere except on Befunge files.

Of course, feel free to modify this to your liking, and look at the scopes contained in the `*.tmLanguage.json` files to see what other Befunge scopes you can customize the theme for.

## Known Issues

* The vertical alignment guides are not centered. I'm sure there must be a way to do this, but if so I haven't been able to think of it!

Find an issue/bug? [Report it](https://github.com/kagof/VSCode-Befunge-syntax-highlighting/issues/new)!

## Contributing

Contributions are encouraged - whether pull requests, bug reports, documentation, or feature requests. Please read, understand, and agree to the [Contributing Guidelines](CONTRIBUTING.md) & [Code of Conduct](CODE_OF_CONDUCT.md) first.

## Source

[Github Repository](https://github.com/kagof/VSCode-Befunge)

## Release Notes

See also the [changelog](CHANGELOG.md).

### 1.2.4

* multiple dependency updates.

### 1.2.3

* Update npm dependencies, addressing several npm vulnerability alerts.

### 1.2.2

* Update npm dependencies addressing (another) low priority npm vulnerability alert.

### 1.2.1

* Update npm dependencies, addressing Github notification of a dependency vulnerability

### 1.2.0

* Restrict colors to 6 digit hex colors as other formats can cause problems, especially with the horizontal lines
* Purge decorations of the old style when color configuration changes, as otherwise it would persist until editor is closed
* Renamed `"befunge.guides.enable"` configuration property to `"befunge.guides.enabled"` to match README and code
* Set some default editor configurations for Befunge-93 and 98 files, including turning off indent guides, line highlight, trailing whitespace trimming, autoindent, and quick suggestions
* Because indent guides are now disabled by default, `"befunge.guides.enabled"` now defaults to `true`

### 1.1.0

* Support for alignment guides for the directional characters

### 1.0.0

* Initial release with support for Befunge-93 and Befunge-98


CHANGELOG.md
# Change Log

All notable changes to the Befunge VS Code extension will be documented in this file.

The format of this changelog is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## 1.2.4 - 2019-07-20

### Changed

* dependecy updates:

  * Bump mocha from 5.2.0 to 6.2.0
  * \[Security\] Bump lodash from 4.17.11 to 4.17.15
  * Bump tslint-config-airbnb from 5.11.0 to 5.11.1
  * Bump typescript from 2.9.2 to 3.5.3
  * Bump @types/mocha from 5.2.5 to 5.2.7
  * Bump tslint from 5.11.0 to 5.18.0

## 1.2.3 - 2019-05-02

### Changed

* long overdue update of dependencies to resolve more npm vulnerability alerts.

## 1.2.2 - 2018-10-09

### Changed

* updated dependencies to address (another) low priority npm dependency vulnerability alert.

## 1.2.1 - 2018-05-01

### Changed

* updated dependencies to address Github notification about [MAID vulnerability in hoek](https://www.cvedetails.com/cve/CVE-2018-3728/) resolved in version `5.0.3`. It appears that version `4.2.1` also fixes this, but to be safe (and to ensure the warning is resolved) all `hoek` dependencies have been updated to `^5.0.3`

## 1.2.0 - 2018-01-22

### Added

* default editor configurations for Befunge-93 and 98 files, including turning off indent guides, line highlight, trailing whitespace trimming, autoindent, and quick suggestions. See the "Contributions" tab in VS Code for details
* restriction on colors: previously accepted any string, now only accepts 6 digit hex colors. This prevents some bugs in rendering, especially in rendering horizontal guides
* extension now renders guides on a change in configuration

### Changed

* fixed bug where guide lines in the old color could still persist after a color change
* `"befunge.guides.enable"` renamed to `"befunge.guides.enabled"`
* `"befunge.guides.enabled"` now defaults to `true`

## 1.1.0 - 2018-01-21

### Added

* Alignment guides from the directional characters (`>`, `v`, `<`, `^`, `?`)

* Configuration to enable/disable alignment guides, and set the color of them per theme.

* Basic testing of the renderer

### Changed

* Changed name from *Befunge Syntax Highlighting* to *Befunge*, to reflect the additional features.

## 1.0.0 (Initial Release) - 2018-01-20

### Added

* Befunge-93 syntax awareness
* Befunge-98 syntax awareness

src/test/index.ts
import * as testRunner from 'vscode/lib/testrunner';

testRunner.configure({
  ui: 'tdd', 		// the TDD UI is being used in whitespace.test.ts (suite, test, etc.)
  useColors: true, // colored output from test results
});

module.exports = testRunner;


src/test/extension.test.ts
//
// Note: This example test is leveraging the Mocha test framework.
// Please refer to their documentation on https://mochajs.org/ for help.
//

// The module 'assert' provides assertion methods from node
import * as assert from 'assert';
import * as typemoq from 'typemoq';

import * as vscode from 'vscode';
import * as extension from '../extension';

suite('Extension Tests', () => {

  const renderer: extension.Renderer = new extension.Renderer();

  test('is whitespace', () => {
    assert.equal(true, renderer.isWhiteSpace(' '));
    assert.equal(true, renderer.isWhiteSpace('\t'));
    assert.equal(true, renderer.isWhiteSpace('\n'));
    assert.equal(false, renderer.isWhiteSpace('f'));
  });

  const mockEd: typemoq.IMock<vscode.TextEditor> = typemoq.Mock.ofType();
  const mockDoc: typemoq.IMock<vscode.TextDocument> = typemoq.Mock.ofType();
  const mockLine: typemoq.IMock<vscode.TextLine> = typemoq.Mock.ofType();
  mockLine.setup(l => l.text).returns(() => '   ');
  mockDoc.setup(d => d.lineCount).returns(() => 3);
  mockDoc.setup(d => d.lineAt(0)).returns(() => mockLine.object);
  mockDoc.setup(d => d.lineAt(1)).returns(() => mockLine.object);
  mockDoc.setup(d => d.lineAt(2)).returns(() => mockLine.object);
  mockEd.setup(e => e.document).returns(() => mockDoc.object);

  test('execute down', () => {
    const ranges: vscode.Range[] = renderer.execDown(new vscode.Position(0, 0), mockEd.object);
    assert.equal(ranges.length, 2, 'number of ranges incorrect');
    assert.equal(ranges[0].isSingleLine, true, 'ranges[0] not single line');
    assert.equal(ranges[1].isSingleLine, true, 'ranges[1] not single line');
    assert.equal(ranges[0].isEqual(new vscode.Range(1, 0, 1, 0)), true, 'ranges[0] did not match');
    assert.equal(ranges[1].isEqual(new vscode.Range(2, 0, 2, 0)), true, 'ranges[1] did not match');
  });

  test('execute up', () => {
    const ranges: vscode.Range[] = renderer.execUp(new vscode.Position(2, 0), mockEd.object);
    assert.equal(ranges.length, 2, 'number of ranges incorrect');
    assert.equal(ranges[0].isSingleLine, true, 'ranges[0] not single line');
    assert.equal(ranges[1].isSingleLine, true, 'ranges[1] not single line');
    assert.equal(ranges[0].isEqual(new vscode.Range(1, 0, 1, 0)), true, 'ranges[0] did not match');
    assert.equal(ranges[1].isEqual(new vscode.Range(0, 0, 0, 0)), true, 'ranges[1] did not match');
  });

  test('execute right', () => {
    const ranges: vscode.Range[] = renderer.execRight(new vscode.Position(0, 0), mockEd.object);
    assert.equal(ranges.length, 1, 'number of ranges incorrect');
    assert.equal(ranges[0].isSingleLine, true, 'not single line');
    assert.equal(ranges[0].isEqual(new vscode.Range(0, 1, 0, 3)), true, 'range did not match');
  });

  test('execute left', () => {
    const ranges: vscode.Range[] = renderer.execLeft(new vscode.Position(0, 3), mockEd.object);
    assert.equal(1, ranges.length, 'number of ranges incorrect');
    assert.equal(true, ranges[0].isSingleLine, 'not single line');
    assert.equal(ranges[0].isEqual(new vscode.Range(0, 0, 0, 3)), true, 'range did not match');
  });

});


src/test/guides/whitespace.test.ts
import * as assert from 'assert';
import * as typemoq from 'typemoq';

import * as vscode from 'vscode';
import * as whitespace from '../../guides/whitespace';

suite('Whitespace Tests', () => {

  test('is whitespace', () => {
    assert.strictEqual(true, whitespace._isWhiteSpace(' '), 'space not whitespace');
    assert.strictEqual(true, whitespace._isWhiteSpace('\t'), 'tab not whitespace');
    assert.strictEqual(true, whitespace._isWhiteSpace('\n'), 'newline not whitespace');
    assert.strictEqual(false, whitespace._isWhiteSpace('f'), 'alpha character not whitespace');
  });

  test('matches', () => {
    assert.strictEqual(true, whitespace._matches('f', 'f'), 'f not in [\'f\']');
    assert.strictEqual(true, whitespace._matches('f', 'f', 'a'), '\'f\' not in [\'f\',\'a\']');
    assert.strictEqual(true, whitespace._matches('f', '', 'f'), '\'f\' not in [\'\',\'f\']');
    assert.strictEqual(false, whitespace._matches('f', 'b', '', 'a'), '\'f\' in [\'b\',\'\',\'a\']');
    assert.strictEqual(false, whitespace._matches('f'), '\'f\' in empty list');
  });

  // sets up the mock VS Code objects
  const mockEd: typemoq.IMock<vscode.TextEditor> = setupMock(
    '?   ?',
    '     ',
    '  ? ', // missing 1 char
    '     ',
    '?   ?',
  );

  test('find connected whitespace down', () => {
    const ranges: vscode.Range[] = whitespace._findConnectedWhitespaceDown(new vscode.Position(0, 0), mockEd.object, false);
    assert.strictEqual(ranges.length, 3, 'number of ranges incorrect');
    assert.strictEqual(ranges[0].isSingleLine, true, 'ranges[0] not single line');
    assert.strictEqual(ranges[1].isSingleLine, true, 'ranges[1] not single line');
    assert.strictEqual(ranges[2].isSingleLine, true, 'ranges[2] not single line');
    assert.strictEqual(ranges[0].isEqual(new vscode.Range(1, 0, 1, 0)), true, 'ranges[0] did not match');
    assert.strictEqual(ranges[1].isEqual(new vscode.Range(2, 0, 2, 0)), true, 'ranges[1] did not match');
    assert.strictEqual(ranges[2].isEqual(new vscode.Range(3, 0, 3, 0)), true, 'ranges[1] did not match');
  });

  test('find connected whitespace down none found', () => {
    const ranges: vscode.Range[] = whitespace._findConnectedWhitespaceDown(new vscode.Position(4, 0), mockEd.object, false);
    assert.strictEqual(ranges.length, 0, 'number of ranges incorrect');
  });

  test('find connected whitespace down with gap', () => {
    const ranges: vscode.Range[] = whitespace._findConnectedWhitespaceDown(new vscode.Position(0, 4), mockEd.object, false);
    assert.strictEqual(ranges.length, 2, 'number of ranges incorrect');
    assert.strictEqual(ranges[0].isSingleLine, true, 'ranges[0] not single line');
    assert.strictEqual(ranges[1].isSingleLine, true, 'ranges[1] not single line');
    assert.strictEqual(ranges[0].isEqual(new vscode.Range(1, 4, 1, 4)), true, 'ranges[0] did not match');
    assert.strictEqual(ranges[1].isEqual(new vscode.Range(3, 4, 3, 4)), true, 'ranges[1] did not match');
  });

  test('find connected whitespace down hits edge', () => {
    const ranges: vscode.Range[] = whitespace._findConnectedWhitespaceDown(new vscode.Position(2, 2), mockEd.object, false);
    assert.strictEqual(ranges.length, 2, 'number of ranges incorrect');
    assert.strictEqual(ranges[0].isSingleLine, true, 'ranges[0] not single line');
    assert.strictEqual(ranges[1].isSingleLine, true, 'ranges[1] not single line');
    assert.strictEqual(ranges[0].isEqual(new vscode.Range(3, 2, 3, 2)), true, 'ranges[0] did not match');
    assert.strictEqual(ranges[1].isEqual(new vscode.Range(4, 2, 4, 2)), true, 'ranges[1] did not match');
  });

  test('find connected whitespace down with wrap', () => {
    const ranges: vscode.Range[] = whitespace._findConnectedWhitespaceDown(new vscode.Position(2, 2), mockEd.object, true);
    assert.strictEqual(ranges.length, 4, 'number of ranges incorrect');
    assert.strictEqual(ranges[0].isSingleLine, true, 'ranges[0] not single line');
    assert.strictEqual(ranges[1].isSingleLine, true, 'ranges[1] not single line');
    assert.strictEqual(ranges[2].isSingleLine, true, 'ranges[2] not single line');
    assert.strictEqual(ranges[3].isSingleLine, true, 'ranges[3] not single line');
    assert.strictEqual(ranges[0].isEqual(new vscode.Range(3, 2, 3, 2)), true, 'ranges[0] did not match');
    assert.strictEqual(ranges[1].isEqual(new vscode.Range(4, 2, 4, 2)), true, 'ranges[1] did not match');
    assert.strictEqual(ranges[2].isEqual(new vscode.Range(0, 2, 0, 2)), true, 'ranges[2] did not match');
    assert.strictEqual(ranges[3].isEqual(new vscode.Range(1, 2, 1, 2)), true, 'ranges[3] did not match');
  });

  test('find connected whitespace up', () => {
    const ranges: vscode.Range[] = whitespace._findConnectedWhitespaceUp(new vscode.Position(4, 0), mockEd.object, false);
    assert.strictEqual(ranges.length, 3, 'number of ranges incorrect');
    assert.strictEqual(ranges[0].isSingleLine, true, 'ranges[0] not single line');
    assert.strictEqual(ranges[1].isSingleLine, true, 'ranges[1] not single line');
    assert.strictEqual(ranges[2].isSingleLine, true, 'ranges[2] not single line');
    assert.strictEqual(ranges[0].isEqual(new vscode.Range(3, 0, 3, 0)), true, 'ranges[0] did not match');
    assert.strictEqual(ranges[1].isEqual(new vscode.Range(2, 0, 2, 0)), true, 'ranges[1] did not match');
    assert.strictEqual(ranges[2].isEqual(new vscode.Range(1, 0, 1, 0)), true, 'ranges[1] did not match');
  });

  test('find connected whitespace up none found', () => {
    const ranges: vscode.Range[] = whitespace._findConnectedWhitespaceUp(new vscode.Position(0, 0), mockEd.object, false);
    assert.strictEqual(ranges.length, 0, 'number of ranges incorrect');
  });

  test('find connected whitespace up with gap', () => {
    const ranges: vscode.Range[] = whitespace._findConnectedWhitespaceUp(new vscode.Position(4, 4), mockEd.object, false);
    assert.strictEqual(ranges.length, 2, 'number of ranges incorrect');
    assert.strictEqual(ranges[0].isSingleLine, true, 'ranges[0] not single line');
    assert.strictEqual(ranges[1].isSingleLine, true, 'ranges[1] not single line');
    assert.strictEqual(ranges[0].isEqual(new vscode.Range(3, 4, 3, 4)), true, 'ranges[0] did not match');
    assert.strictEqual(ranges[1].isEqual(new vscode.Range(1, 4, 1, 4)), true, 'ranges[1] did not match');
  });

  test('find connected whitespace up hits edge', () => {
    const ranges: vscode.Range[] = whitespace._findConnectedWhitespaceUp(new vscode.Position(2, 2), mockEd.object, false);
    assert.strictEqual(ranges.length, 2, 'number of ranges incorrect');
    assert.strictEqual(ranges[0].isSingleLine, true, 'ranges[0] not single line');
    assert.strictEqual(ranges[1].isSingleLine, true, 'ranges[1] not single line');
    assert.strictEqual(ranges[0].isEqual(new vscode.Range(1, 2, 1, 2)), true, 'ranges[0] did not match');
    assert.strictEqual(ranges[1].isEqual(new vscode.Range(0, 2, 0, 2)), true, 'ranges[1] did not match');
  });

  test('find connected whitespace up with wrap', () => {
    const ranges: vscode.Range[] = whitespace._findConnectedWhitespaceUp(new vscode.Position(2, 2), mockEd.object, true);
    assert.strictEqual(ranges.length, 4, 'number of ranges incorrect');
    assert.strictEqual(ranges[0].isSingleLine, true, 'ranges[0] not single line');
    assert.strictEqual(ranges[1].isSingleLine, true, 'ranges[1] not single line');
    assert.strictEqual(ranges[2].isSingleLine, true, 'ranges[2] not single line');
    assert.strictEqual(ranges[3].isSingleLine, true, 'ranges[3] not single line');
    assert.strictEqual(ranges[0].isEqual(new vscode.Range(1, 2, 1, 2)), true, 'ranges[0] did not match');
    assert.strictEqual(ranges[1].isEqual(new vscode.Range(0, 2, 0, 2)), true, 'ranges[1] did not match');
    assert.strictEqual(ranges[2].isEqual(new vscode.Range(4, 2, 4, 2)), true, 'ranges[2] did not match');
    assert.strictEqual(ranges[3].isEqual(new vscode.Range(3, 2, 3, 2)), true, 'ranges[3] did not match');
  });

  test('find connected whitespace right', () => {
    const ranges: vscode.Range[] = whitespace._findConnectedWhitespaceRight(new vscode.Position(0, 0), mockEd.object, false);
    assert.strictEqual(ranges.length, 1, 'number of ranges incorrect');
    assert.strictEqual(ranges[0].isSingleLine, true, 'not single line');
    assert.strictEqual(ranges[0].isEqual(new vscode.Range(0, 1, 0, 4)), true, 'range did not match');
  });

  test('find connected whitespace right none found', () => {
    const ranges: vscode.Range[] = whitespace._findConnectedWhitespaceRight(new vscode.Position(0, 4), mockEd.object, false);
    assert.strictEqual(ranges.length, 0, 'number of ranges incorrect');
  });

  test('find connected whitespace right hits edge', () => {
    const ranges: vscode.Range[] = whitespace._findConnectedWhitespaceRight(new vscode.Position(2, 2), mockEd.object, false);
    assert.strictEqual(ranges.length, 1, 'number of ranges incorrect');
    assert.strictEqual(ranges[0].isSingleLine, true, 'not single line');
    assert.strictEqual(ranges[0].isEqual(new vscode.Range(2, 3, 2, 4)), true, 'range did not match');
  });

  test('find connected whitespace right with wrap', () => {
    const ranges: vscode.Range[] = whitespace._findConnectedWhitespaceRight(new vscode.Position(2, 2), mockEd.object, true);
    assert.strictEqual(ranges.length, 2, 'number of ranges incorrect');
    assert.strictEqual(ranges[0].isSingleLine, true, 'ranges[0] not single line');
    assert.strictEqual(ranges[1].isSingleLine, true, 'ranges[1] not single line');
    assert.strictEqual(ranges[0].isEqual(new vscode.Range(2, 3, 2, 4)), true, 'ranges[0] did not match');
    assert.strictEqual(ranges[1].isEqual(new vscode.Range(2, 0, 2, 2)), true, 'ranges[1] did not match');
  });

  test('find connected whitespace left', () => {
    const ranges: vscode.Range[] = whitespace._findConnectedWhitespaceLeft(new vscode.Position(0, 4), mockEd.object, false);
    assert.strictEqual(ranges.length, 1, 'number of ranges incorrect');
    assert.strictEqual(ranges[0].isSingleLine, true, 'not single line');
    assert.strictEqual(ranges[0].isEqual(new vscode.Range(0, 1, 0, 4)), true, 'range did not match');
  });

  test('find connected whitespace left none found', () => {
    const ranges: vscode.Range[] = whitespace._findConnectedWhitespaceLeft(new vscode.Position(0, 0), mockEd.object, false);
    assert.strictEqual(ranges.length, 0, 'number of ranges incorrect');
  });

  test('find connected whitespace left hits edge', () => {
    const ranges: vscode.Range[] = whitespace._findConnectedWhitespaceLeft(new vscode.Position(2, 2), mockEd.object, false);
    assert.strictEqual(ranges.length, 1, 'number of ranges incorrect');
    assert.strictEqual(ranges[0].isSingleLine, true, 'not single line');
    assert.strictEqual(ranges[0].isEqual(new vscode.Range(2, 0, 2, 2)), true, 'range did not match');
  });

  test('find connected whitespace left with wrap', () => {
    const ranges: vscode.Range[] = whitespace._findConnectedWhitespaceLeft(new vscode.Position(2, 2), mockEd.object, true);
    assert.strictEqual(ranges.length, 2, 'number of ranges incorrect');
    assert.strictEqual(ranges[0].isSingleLine, true, 'ranges[0] not single line');
    assert.strictEqual(ranges[1].isSingleLine, true, 'ranges[1] not single line');
    assert.strictEqual(ranges[0].isEqual(new vscode.Range(2, 0, 2, 2)), true, 'ranges[0] did not match');
    assert.strictEqual(ranges[1].isEqual(new vscode.Range(2, 3, 2, 4)), true, 'ranges[1] did not match');
  });

  // used to test a corner case in vertical wrapping
  const edVertWrapCorner: typemoq.IMock<vscode.TextEditor> = setupMock(
    '',
    ' ',
    '?',
    ' ',
    '',
  );

  test('find connected whitespace down with wrap edge is not present still wraps', () => {
    const rs: vscode.Range[] = whitespace._findConnectedWhitespaceDown(new vscode.Position(2, 0), edVertWrapCorner.object, true);
    assert.strictEqual(rs.length, 2, 'number of ranges incorrect');
    assert.strictEqual(rs[0].isSingleLine, true, 'ranges[0] not single line');
    assert.strictEqual(rs[1].isSingleLine, true, 'ranges[1] not single line');
    assert.strictEqual(rs[0].isEqual(new vscode.Range(3, 0, 3, 0)), true, 'ranges[0] did not match');
    assert.strictEqual(rs[1].isEqual(new vscode.Range(1, 0, 1, 0)), true, 'ranges[1] did not match');
  });

  test('find connected whitespace up with wrap edge is not present still wraps', () => {
    const rs: vscode.Range[] = whitespace._findConnectedWhitespaceUp(new vscode.Position(2, 0), edVertWrapCorner.object, true);
    assert.strictEqual(rs.length, 2, 'number of ranges incorrect');
    assert.strictEqual(rs[0].isSingleLine, true, 'ranges[0] not single line');
    assert.strictEqual(rs[1].isSingleLine, true, 'ranges[1] not single line');
    assert.strictEqual(rs[0].isEqual(new vscode.Range(1, 0, 1, 0)), true, 'ranges[0] did not match');
    assert.strictEqual(rs[1].isEqual(new vscode.Range(3, 0, 3, 0)), true, 'ranges[1] did not match');
  });

});

function setupMock(...rows: string[]): typemoq.IMock<vscode.TextEditor> {
  const mockEd: typemoq.IMock<vscode.TextEditor> = typemoq.Mock.ofType();
  const mockDoc: typemoq.IMock<vscode.TextDocument> = typemoq.Mock.ofType();
  mockDoc.setup(d => d.lineCount).returns(() => 5);
  rows.forEach((row, index) => {
    const mockLine: typemoq.IMock<vscode.TextLine> = typemoq.Mock.ofType();
    mockLine.setup(l => l.text).returns(() => row);
    mockDoc.setup(d => d.lineAt(index)).returns(() => mockLine.object);
  });
  mockEd.setup(e => e.document).returns(() => mockDoc.object);
  return mockEd;
}


src/guides/renderer.ts
import * as vscode from 'vscode';
import * as whitespace from './whitespace';
import { Configurer } from './configurer';

/**
 * renders the guides.
 */
export class Renderer {
  private configurer: Configurer;

  shouldWrap: boolean = false;

  verticalDecoration: vscode.TextEditorDecorationType;
  horizontalDecoration: vscode.TextEditorDecorationType;

  constructor() {
    this.configurer = new Configurer();
  }

  /**
   * renders the guides, if guides are enabled.
   * @param editor the current text editor
   */
  render(editor: vscode.TextEditor): void {
    if (!this.configurer.isEnabled()) {
      return;
    }
    this.configurer.refresh(this); // refreshes the renderer's configuration

    const ranges: whitespace.HorizontalAndVerticalRanges = whitespace.findWhitespaceRanges(editor, this.shouldWrap);

    editor.setDecorations(this.verticalDecoration, ranges.verticalRanges);
    editor.setDecorations(this.horizontalDecoration, ranges.horizontalRanges);
  }

}


src/guides/configurer.ts
import * as vscode from 'vscode';
import { Renderer } from './renderer';

export const CONFIGURATION_NAME = 'befunge.guides';

/**
 * configures the renderer.
 */
export class Configurer {
  private colorDark: String;
  private colorLight: String;
  private config: vscode.WorkspaceConfiguration;

  /**
   * Returns whether or not alignment guides are enabled in the workspace configuration.
   */
  isEnabled(): boolean {
    return this.forceRefreshConfig().enabled;
  }

  /**
   * Rereads the configuration, purging old decorations and reloading styles as needed.
   * @param renderer the renderer to refresh the configuration of
   * @param forceRefreshConfig whether to force a configuration refresh or not
   */
  refresh(renderer: Renderer, forceRefreshConfig: boolean = false): void {
    const cfg: vscode.WorkspaceConfiguration = forceRefreshConfig ? this.forceRefreshConfig() : this.getConfig();
    renderer.shouldWrap = cfg.wrap;
    if (this.colorDark !== cfg.color.dark || this.colorLight !== cfg.color.light) {
      renderer.verticalDecoration && renderer.verticalDecoration.dispose(); // erase all vertical decorations in old style
      renderer.horizontalDecoration && renderer.horizontalDecoration.dispose(); // erase all horizontal decorations in old style
      this.loadDecorations(renderer, cfg);
    }
  }

  /**
   * retrieves the current workspace configuration.
   */
  private getConfig(): vscode.WorkspaceConfiguration {
    if (this.config == null) {
      this.forceRefreshConfig();
    }
    return this.config;
  }

  /**
   * forces updating the current workspace configuration.
   */
  private forceRefreshConfig(): vscode.WorkspaceConfiguration {
    this.config = vscode.workspace.getConfiguration(CONFIGURATION_NAME);
    return this.config;
  }

  /**
   * Refreshes the decorations. To be called when a color configuration has been changed.
   * @param renderer the renderer to update, if needed
   * @param cfg the configuration to load the decorations from
   */
  private loadDecorations(renderer: Renderer, cfg: vscode.WorkspaceConfiguration): void {
    this.colorDark = cfg.color.dark;
    this.colorLight = cfg.color.light;
    const verticalConfig = {
      dark: {
        outlineWidth: '1px',
        outlineColor: this.colorDark,
        outlineStyle: 'solid',
      },
      light: {
        outlineWidth: '1px',
        outlineColor: this.colorLight,
        outlineStyle: 'solid',
      },
    };
    const horizontalConfig = {
      dark: {
        textDecoration: `line-through ${this.colorDark}`,
      },
      light: {
        textDecoration: `line-through ${this.colorLight}`,
      },
    };
    renderer.verticalDecoration = vscode.window.createTextEditorDecorationType(verticalConfig);
    renderer.horizontalDecoration = vscode.window.createTextEditorDecorationType(horizontalConfig);
  }
}


src/guides/listener.ts
import * as vscode from 'vscode';
import { Renderer } from './renderer';
import { CONFIGURATION_NAME } from './configurer';

/**
 * listens to events that affect the Befunge guides.
 */
export class Listener {
  private renderer: Renderer;
  private editor: vscode.TextEditor;

  constructor (private context: vscode.ExtensionContext) {
    this.renderer = new Renderer();
  }

  /**
   * begin listening for changes.
   */
  listen(): void {
    this.editor = vscode.window.activeTextEditor;
    if (this.isBefungeDoc(this.editor)) {
      this.renderer.render(this.editor);
    }
    this.listenToDidChangeActiveTextEditor();
    this.listenToDidChangeTextDocument();
    this.listenToDidChangeConfiguration();
  }

  /**
   * Begins listening for changes to the active text editor.
   */
  private listenToDidChangeActiveTextEditor(): void {
    vscode.window.onDidChangeActiveTextEditor(
      (ed) => {
        this.editor = ed;
        if (this.isBefungeDoc(this.editor)) {
          this.renderer.render(this.editor);
        }
      },
      null, this.context.subscriptions);
  }

  /**
   * Begins listening for changes to the active text document.
   */
  private listenToDidChangeTextDocument(): void {
    vscode.workspace.onDidChangeTextDocument(
      (event) => {
        if (this.editor && event.document === this.editor.document && this.isBefungeDoc(this.editor)) {
          this.renderer.render(this.editor);
        }
      },
      null, this.context.subscriptions);
  }

  /**
   * Begins listening for changes to the befunge guides configuration.
   */
  private listenToDidChangeConfiguration(): void {
    vscode.workspace.onDidChangeConfiguration(
      (event: vscode.ConfigurationChangeEvent) => {
        if (event.affectsConfiguration(CONFIGURATION_NAME) && this.isBefungeDoc(this.editor)) {
          this.renderer.render(this.editor);
        }
      },
      null, this.context.subscriptions);
  }

  /**
   * Determines if the language being used by the editor is Befunge (93 or 98)
   * @param editor the current text editor
   */
  private isBefungeDoc(editor?: vscode.TextEditor): boolean {
    return (editor && editor.document &&
      (editor.document.languageId === 'befunge' || editor.document.languageId === 'befunge98'));
  }
}
